In [46]:
import geocoder

In [5]:
import numpy as np
import pandas as pd
import scipy as sp
import scipy.stats as stats
from pandas import DataFrame, Series
import matplotlib.pyplot as plt
import matplotlib
import matplotlib as mpl
import seaborn as sns
import pandas as pd
# Numpy options
np.random.seed(78)
# Pandas options
pd.set_option("display.max_rows", 100)
pd.set_option("display.max_columns", 400)
# Matplotlib options
matplotlib.rcParams['image.cmap'] = 'viridis'
%matplotlib inline
matplotlib.style.use('ggplot')
matplotlib.rcParams['figure.figsize'] = (10,6)

In [45]:
from geopy.geocoders import Nominatim
geolocator = Nominatim()
location = geolocator.geocode("Mercer Street and Canal Street")
print(location.latitude,location.longitude)

GeocoderServiceError: HTTP Error 429: Too Many Requests

'40.7410861'

In [104]:
df = pd.read_csv('aadt_cleaned.csv')

In [105]:
df = df.drop(df.columns[0], axis=1)

In [106]:
df

Roadway Begin Description       Roadway End Description  \
0                          NAVY ST                      BROADWAY   
1                         CANAL ST                      E 8TH ST   
2                      EBBITTS AVE                 GARIBALDI AVE   
3                          4TH AVE                    SECOND AVE   
4                            AVE X                       86TH ST   
5       END 1/95I OLAP WEBSTER AVE                  E FORDHAM RD   
6       END 1/95I OLAP WEBSTER AVE                  E FORDHAM RD   
7                     E FORDHAM RD  RT 907H BRONX RIVER PKWY SOU   
8                     E FORDHAM RD  RT 907H BRONX RIVER PKWY SOU   
9     RT 907H BRONX RIVER PKWY SOU  PELHAM PKWY BOSTON POST RD R   
10    RT 907H BRONX RIVER PKWY SOU  PELHAM PKWY BOSTON POST RD R   
11    PELHAM PKWY BOSTON POST RD R                   GUN HILL RD   
12    PELHAM PKWY BOSTON POST RD R                   GUN HILL RD   
13                     GUN HILL RD             RT 22 PROVOST AVE   
14                     GUN HILL RD             RT 22 PROVOST AVE   
15               RT 22 PROVOST AVE            BRONX/WEST CO LINE   
16               RT 22 PROVOST AVE            BRONX/WEST CO LINE   
17               RT 22 PROVOST AVE            BRONX/WEST CO LINE   
18           ACC BROADWAY 178TH ST                    DYCKMAN ST   
19           ACC BROADWAY 178TH ST                    DYCKMAN ST   
20                      DYCKMAN ST              NY/BRONX CO LINE   
21                      DYCKMAN ST              NY/BRONX CO LINE   
22                NY/BRONX CO LINE            VAN CORTLAND SOUTH   
23                NY/BRONX CO LINE            VAN CORTLAND SOUTH   
24              VAN CORTLAND SOUTH      START 9/9A OLAP H.H. PKY   
25              VAN CORTLAND SOUTH      START 9/9A OLAP H.H. PKY   
26        START 9/9A OLAP H.H. PKY            BRONX/WEST CO LINE   
27        START 9/9A OLAP H.H. PKY            BRONX/WEST CO LINE   
28        START 9/9A OLAP H.H. PKY            BRONX/WEST CO LINE   
29                            RT 1                 EAST 233RD ST   
30                            RT 1                 EAST 233RD ST   
31                   EAST 233RD ST            BRONX/WEST CO LINE   
32                   EAST 233RD ST            BRONX/WEST CO LINE   
33                    HILLSIDE AVE                   JAMAICA AVE   
34                    HILLSIDE AVE                   JAMAICA AVE   
35                     JAMAICA AVE                 HEMPSTEAD AVE   
36                     JAMAICA AVE                 HEMPSTEAD AVE   
37                   HEMPSTEAD AVE             CROSS ISLAND PKWY   
38                   HEMPSTEAD AVE             CROSS ISLAND PKWY   
39               CROSS ISLAND PKWY         QUEENS/NASSAU CO LINE   
40               CROSS ISLAND PKWY         QUEENS/NASSAU CO LINE   
41               CROSS ISLAND PKWY         QUEENS/NASSAU CO LINE   
42             2ND AVE - MANHATTAN             NY/QUEENS CO LINE   
43             2ND AVE - MANHATTAN             NY/QUEENS CO LINE   
44               NY/QUEENS CO LINE                   JACKSON AVE   
45               NY/QUEENS CO LINE                   JACKSON AVE   
46                     JACKSON AVE                278I BQE UNDER   
47                     JACKSON AVE                278I BQE UNDER   
48                  278I BQE UNDER          INT YELLOWSTONE BLVD   
49                  278I BQE UNDER          INT YELLOWSTONE BLVD   
...                            ...                           ...   
6283               WILLIAMSBURG BR                     RODNEY ST   
6284               WILLIAMSBURG BR                     RODNEY ST   
6285                     RODNEY ST                     UNION AVE   
6286                     RODNEY ST                     UNION AVE   
6287                       31ST ST                      135TH PL   
6288                       31ST ST                      135TH PL   
6289                      224TH ST      230TH PL FRAN LEWIS BLVD 

In [142]:
for d in range(5443,len(df)):
    street1 = df.at[d,'Roadway Begin Description']
    street2 = df.at[d,'Road Name']
    street3 = street1+ ' '+street2
    g = geocoder.google(street3)
    if len(g.latlng)>0:
        df.at[d,'latlng'] = str(g.latlng[0]) + ',' + str(g.latlng[1])
#         print((g.latlng))
    else:
        street3 = street2+' '+street1
        g = geocoder.google(street3)
        if len(g.latlng)>0:
            df.at[d,'latlng'] = str(g.latlng[0]) + ',' + str(g.latlng[1])

In [146]:
df.to_csv('latlngData.csv')

In [44]:
import geocoder

In [61]:
g = geocoder.google('Mountain View, CA')
print(g.latlng[0])

<class 'float'>


In [161]:
df2 = pd.read_csv('aadt_gps.csv')

In [162]:
df2
df2 = df2.drop(df2.columns[0], axis=1)

In [163]:
df2

Roadway Begin Description       Roadway End Description  \
0                          NAVY ST                      BROADWAY   
1                         CANAL ST                      E 8TH ST   
2                      EBBITTS AVE                 GARIBALDI AVE   
3                          4TH AVE                    SECOND AVE   
4                            AVE X                       86TH ST   
5       END 1/95I OLAP WEBSTER AVE                  E FORDHAM RD   
6       END 1/95I OLAP WEBSTER AVE                  E FORDHAM RD   
7                     E FORDHAM RD  RT 907H BRONX RIVER PKWY SOU   
8                     E FORDHAM RD  RT 907H BRONX RIVER PKWY SOU   
9     RT 907H BRONX RIVER PKWY SOU  PELHAM PKWY BOSTON POST RD R   
10    RT 907H BRONX RIVER PKWY SOU  PELHAM PKWY BOSTON POST RD R   
11    PELHAM PKWY BOSTON POST RD R                   GUN HILL RD   
12    PELHAM PKWY BOSTON POST RD R                   GUN HILL RD   
13                     GUN HILL RD             RT 22 PROVOST AVE   
14                     GUN HILL RD             RT 22 PROVOST AVE   
15               RT 22 PROVOST AVE            BRONX/WEST CO LINE   
16               RT 22 PROVOST AVE            BRONX/WEST CO LINE   
17               RT 22 PROVOST AVE            BRONX/WEST CO LINE   
18           ACC BROADWAY 178TH ST                    DYCKMAN ST   
19           ACC BROADWAY 178TH ST                    DYCKMAN ST   
20                      DYCKMAN ST              NY/BRONX CO LINE   
21                      DYCKMAN ST              NY/BRONX CO LINE   
22                NY/BRONX CO LINE            VAN CORTLAND SOUTH   
23                NY/BRONX CO LINE            VAN CORTLAND SOUTH   
24              VAN CORTLAND SOUTH      START 9/9A OLAP H.H. PKY   
25              VAN CORTLAND SOUTH      START 9/9A OLAP H.H. PKY   
26        START 9/9A OLAP H.H. PKY            BRONX/WEST CO LINE   
27        START 9/9A OLAP H.H. PKY            BRONX/WEST CO LINE   
28        START 9/9A OLAP H.H. PKY            BRONX/WEST CO LINE   
29                            RT 1                 EAST 233RD ST   
30                            RT 1                 EAST 233RD ST   
31                   EAST 233RD ST            BRONX/WEST CO LINE   
32                   EAST 233RD ST            BRONX/WEST CO LINE   
33                    HILLSIDE AVE                   JAMAICA AVE   
34                    HILLSIDE AVE                   JAMAICA AVE   
35                     JAMAICA AVE                 HEMPSTEAD AVE   
36                     JAMAICA AVE                 HEMPSTEAD AVE   
37                   HEMPSTEAD AVE             CROSS ISLAND PKWY   
38                   HEMPSTEAD AVE             CROSS ISLAND PKWY   
39               CROSS ISLAND PKWY         QUEENS/NASSAU CO LINE   
40               CROSS ISLAND PKWY         QUEENS/NASSAU CO LINE   
41               CROSS ISLAND PKWY         QUEENS/NASSAU CO LINE   
42             2ND AVE - MANHATTAN             NY/QUEENS CO LINE   
43             2ND AVE - MANHATTAN             NY/QUEENS CO LINE   
44               NY/QUEENS CO LINE                   JACKSON AVE   
45               NY/QUEENS CO LINE                   JACKSON AVE   
46                     JACKSON AVE                278I BQE UNDER   
47                     JACKSON AVE                278I BQE UNDER   
48                  278I BQE UNDER          INT YELLOWSTONE BLVD   
49                  278I BQE UNDER          INT YELLOWSTONE BLVD   
...                            ...                           ...   
6283               WILLIAMSBURG BR                     RODNEY ST   
6284               WILLIAMSBURG BR                     RODNEY ST   
6285                     RODNEY ST                     UNION AVE   
6286                     RODNEY ST                     UNION AVE   
6287                       31ST ST                      135TH PL   
6288                       31ST ST                      135TH PL   
6289                      224TH ST      230TH PL FRAN LEWIS BLVD 

In [165]:
df2=df2[pd.notnull(df2['latlng'])]
df2

Roadway Begin Description       Roadway End Description  \
0                          NAVY ST                      BROADWAY   
1                         CANAL ST                      E 8TH ST   
2                      EBBITTS AVE                 GARIBALDI AVE   
3                          4TH AVE                    SECOND AVE   
4                            AVE X                       86TH ST   
7                     E FORDHAM RD  RT 907H BRONX RIVER PKWY SOU   
11    PELHAM PKWY BOSTON POST RD R                   GUN HILL RD   
12    PELHAM PKWY BOSTON POST RD R                   GUN HILL RD   
13                     GUN HILL RD             RT 22 PROVOST AVE   
14                     GUN HILL RD             RT 22 PROVOST AVE   
20                      DYCKMAN ST              NY/BRONX CO LINE   
21                      DYCKMAN ST              NY/BRONX CO LINE   
24              VAN CORTLAND SOUTH      START 9/9A OLAP H.H. PKY   
25              VAN CORTLAND SOUTH      START 9/9A OLAP H.H. PKY   
29                            RT 1                 EAST 233RD ST   
30                            RT 1                 EAST 233RD ST   
31                   EAST 233RD ST            BRONX/WEST CO LINE   
32                   EAST 233RD ST            BRONX/WEST CO LINE   
33                    HILLSIDE AVE                   JAMAICA AVE   
34                    HILLSIDE AVE                   JAMAICA AVE   
35                     JAMAICA AVE                 HEMPSTEAD AVE   
36                     JAMAICA AVE                 HEMPSTEAD AVE   
37                   HEMPSTEAD AVE             CROSS ISLAND PKWY   
38                   HEMPSTEAD AVE             CROSS ISLAND PKWY   
39               CROSS ISLAND PKWY         QUEENS/NASSAU CO LINE   
40               CROSS ISLAND PKWY         QUEENS/NASSAU CO LINE   
41               CROSS ISLAND PKWY         QUEENS/NASSAU CO LINE   
42             2ND AVE - MANHATTAN             NY/QUEENS CO LINE   
43             2ND AVE - MANHATTAN             NY/QUEENS CO LINE   
44               NY/QUEENS CO LINE                   JACKSON AVE   
45               NY/QUEENS CO LINE                   JACKSON AVE   
46                     JACKSON AVE                278I BQE UNDER   
47                     JACKSON AVE                278I BQE UNDER   
50            INT YELLOWSTONE BLVD  RT 908B INT UNION TPKE (INTE   
51            INT YELLOWSTONE BLVD  RT 908B INT UNION TPKE (INTE   
56                    HILLSIDE AVE            FRANCIS LEWIS BLVD   
57                    HILLSIDE AVE            FRANCIS LEWIS BLVD   
58              FRANCIS LEWIS BLVD           RT 25B BRADDOCK AVE   
59              FRANCIS LEWIS BLVD           RT 25B BRADDOCK AVE   
60             RT 25B BRADDOCK AVE              CROSS ISLAND EXP   
61             RT 25B BRADDOCK AVE              CROSS ISLAND EXP   
62                CROSS ISLAND EXP         QUEENS/NASSAU CO LINE   
63                CROSS ISLAND EXP         QUEENS/NASSAU CO LINE   
64                       CATON AVE              CONEY ISLAND AVE   
65                       CATON AVE              CONEY ISLAND AVE   
66                CONEY ISLAND AVE                   ST PAULS PL   
67                CONEY ISLAND AVE                   ST PAULS PL   
68                     ST PAULS PL                  FLATBUSH AVE   
69                     ST PAULS PL                  FLATBUSH AVE   
70                    FLATBUSH AVE            JCT LINDEN / CATON   
...                            ...                           ...   
6271                     135TH AVE                     109TH AVE   
6272                     135TH AVE                     109TH AVE   
6273                   JAMAICA AVE                      HILLSIDE   
6274                   JAMAICA AVE                      HILLSIDE   
6275                      178TH PL                  FARMERS BLVD   
6276                      178TH PL                  FARMERS BLVD   
6277                      226TH ST                      229TH ST 

In [166]:
df2 = df2[df2['latlng']>str(40)]

In [167]:
df2

Roadway Begin Description       Roadway End Description  \
0                          NAVY ST                      BROADWAY   
1                         CANAL ST                      E 8TH ST   
2                      EBBITTS AVE                 GARIBALDI AVE   
3                          4TH AVE                    SECOND AVE   
7                     E FORDHAM RD  RT 907H BRONX RIVER PKWY SOU   
11    PELHAM PKWY BOSTON POST RD R                   GUN HILL RD   
12    PELHAM PKWY BOSTON POST RD R                   GUN HILL RD   
13                     GUN HILL RD             RT 22 PROVOST AVE   
14                     GUN HILL RD             RT 22 PROVOST AVE   
20                      DYCKMAN ST              NY/BRONX CO LINE   
21                      DYCKMAN ST              NY/BRONX CO LINE   
24              VAN CORTLAND SOUTH      START 9/9A OLAP H.H. PKY   
25              VAN CORTLAND SOUTH      START 9/9A OLAP H.H. PKY   
29                            RT 1                 EAST 233RD ST   
30                            RT 1                 EAST 233RD ST   
31                   EAST 233RD ST            BRONX/WEST CO LINE   
32                   EAST 233RD ST            BRONX/WEST CO LINE   
33                    HILLSIDE AVE                   JAMAICA AVE   
34                    HILLSIDE AVE                   JAMAICA AVE   
35                     JAMAICA AVE                 HEMPSTEAD AVE   
36                     JAMAICA AVE                 HEMPSTEAD AVE   
37                   HEMPSTEAD AVE             CROSS ISLAND PKWY   
38                   HEMPSTEAD AVE             CROSS ISLAND PKWY   
39               CROSS ISLAND PKWY         QUEENS/NASSAU CO LINE   
40               CROSS ISLAND PKWY         QUEENS/NASSAU CO LINE   
41               CROSS ISLAND PKWY         QUEENS/NASSAU CO LINE   
42             2ND AVE - MANHATTAN             NY/QUEENS CO LINE   
43             2ND AVE - MANHATTAN             NY/QUEENS CO LINE   
44               NY/QUEENS CO LINE                   JACKSON AVE   
45               NY/QUEENS CO LINE                   JACKSON AVE   
46                     JACKSON AVE                278I BQE UNDER   
47                     JACKSON AVE                278I BQE UNDER   
50            INT YELLOWSTONE BLVD  RT 908B INT UNION TPKE (INTE   
51            INT YELLOWSTONE BLVD  RT 908B INT UNION TPKE (INTE   
56                    HILLSIDE AVE            FRANCIS LEWIS BLVD   
57                    HILLSIDE AVE            FRANCIS LEWIS BLVD   
58              FRANCIS LEWIS BLVD           RT 25B BRADDOCK AVE   
59              FRANCIS LEWIS BLVD           RT 25B BRADDOCK AVE   
60             RT 25B BRADDOCK AVE              CROSS ISLAND EXP   
61             RT 25B BRADDOCK AVE              CROSS ISLAND EXP   
62                CROSS ISLAND EXP         QUEENS/NASSAU CO LINE   
63                CROSS ISLAND EXP         QUEENS/NASSAU CO LINE   
66                CONEY ISLAND AVE                   ST PAULS PL   
67                CONEY ISLAND AVE                   ST PAULS PL   
68                     ST PAULS PL                  FLATBUSH AVE   
69                     ST PAULS PL                  FLATBUSH AVE   
70                    FLATBUSH AVE            JCT LINDEN / CATON   
71                    FLATBUSH AVE            JCT LINDEN / CATON   
72              JCT LINDEN / CATON                 KINGS HIGHWAY   
73              JCT LINDEN / CATON                 KINGS HIGHWAY   
...                            ...                           ...   
6253                  BAISLEY BLVD              SPRINGFIELD BLVD   
6254                  BAISLEY BLVD              SPRINGFIELD BLVD   
6261                 N CONDUIT AVE                 ROCKAWAY BLVD   
6262                 N CONDUIT AVE                 ROCKAWAY BLVD   
6263                       7TH AVE                      20TH AVE   
6264                       7TH AVE                      20TH AVE   
6265              SPRINGFIELD BLVD                      237TH ST 

In [168]:
df2 = df2.reset_index(drop=True)

In [169]:
df2

Roadway Begin Description       Roadway End Description  \
0                          NAVY ST                      BROADWAY   
1                         CANAL ST                      E 8TH ST   
2                      EBBITTS AVE                 GARIBALDI AVE   
3                          4TH AVE                    SECOND AVE   
4                     E FORDHAM RD  RT 907H BRONX RIVER PKWY SOU   
5     PELHAM PKWY BOSTON POST RD R                   GUN HILL RD   
6     PELHAM PKWY BOSTON POST RD R                   GUN HILL RD   
7                      GUN HILL RD             RT 22 PROVOST AVE   
8                      GUN HILL RD             RT 22 PROVOST AVE   
9                       DYCKMAN ST              NY/BRONX CO LINE   
10                      DYCKMAN ST              NY/BRONX CO LINE   
11              VAN CORTLAND SOUTH      START 9/9A OLAP H.H. PKY   
12              VAN CORTLAND SOUTH      START 9/9A OLAP H.H. PKY   
13                            RT 1                 EAST 233RD ST   
14                            RT 1                 EAST 233RD ST   
15                   EAST 233RD ST            BRONX/WEST CO LINE   
16                   EAST 233RD ST            BRONX/WEST CO LINE   
17                    HILLSIDE AVE                   JAMAICA AVE   
18                    HILLSIDE AVE                   JAMAICA AVE   
19                     JAMAICA AVE                 HEMPSTEAD AVE   
20                     JAMAICA AVE                 HEMPSTEAD AVE   
21                   HEMPSTEAD AVE             CROSS ISLAND PKWY   
22                   HEMPSTEAD AVE             CROSS ISLAND PKWY   
23               CROSS ISLAND PKWY         QUEENS/NASSAU CO LINE   
24               CROSS ISLAND PKWY         QUEENS/NASSAU CO LINE   
25               CROSS ISLAND PKWY         QUEENS/NASSAU CO LINE   
26             2ND AVE - MANHATTAN             NY/QUEENS CO LINE   
27             2ND AVE - MANHATTAN             NY/QUEENS CO LINE   
28               NY/QUEENS CO LINE                   JACKSON AVE   
29               NY/QUEENS CO LINE                   JACKSON AVE   
30                     JACKSON AVE                278I BQE UNDER   
31                     JACKSON AVE                278I BQE UNDER   
32            INT YELLOWSTONE BLVD  RT 908B INT UNION TPKE (INTE   
33            INT YELLOWSTONE BLVD  RT 908B INT UNION TPKE (INTE   
34                    HILLSIDE AVE            FRANCIS LEWIS BLVD   
35                    HILLSIDE AVE            FRANCIS LEWIS BLVD   
36              FRANCIS LEWIS BLVD           RT 25B BRADDOCK AVE   
37              FRANCIS LEWIS BLVD           RT 25B BRADDOCK AVE   
38             RT 25B BRADDOCK AVE              CROSS ISLAND EXP   
39             RT 25B BRADDOCK AVE              CROSS ISLAND EXP   
40                CROSS ISLAND EXP         QUEENS/NASSAU CO LINE   
41                CROSS ISLAND EXP         QUEENS/NASSAU CO LINE   
42                CONEY ISLAND AVE                   ST PAULS PL   
43                CONEY ISLAND AVE                   ST PAULS PL   
44                     ST PAULS PL                  FLATBUSH AVE   
45                     ST PAULS PL                  FLATBUSH AVE   
46                    FLATBUSH AVE            JCT LINDEN / CATON   
47                    FLATBUSH AVE            JCT LINDEN / CATON   
48              JCT LINDEN / CATON                 KINGS HIGHWAY   
49              JCT LINDEN / CATON                 KINGS HIGHWAY   
...                            ...                           ...   
4450                  BAISLEY BLVD              SPRINGFIELD BLVD   
4451                  BAISLEY BLVD              SPRINGFIELD BLVD   
4452                 N CONDUIT AVE                 ROCKAWAY BLVD   
4453                 N CONDUIT AVE                 ROCKAWAY BLVD   
4454                       7TH AVE                      20TH AVE   
4455                       7TH AVE                      20TH AVE   
4456              SPRINGFIELD BLVD                      237TH ST 

In [171]:
df2.to_csv('aadt_gps_cleaned.csv')

In [208]:
g = geocoder.google([45.15,-75.14], method='reverse')

In [209]:
g.debug()

{
    "postal_code": {},
    "location": {},
    "locality": {},
    "sublocality": {},
    "administrative_area_level_2": {},
    "northeast": {},
    "country": {},
    "subpremise": {},
    "neighborhood": {},
    "administrative_area_level_1": {},
    "status": "INVALID_REQUEST",
    "results": [],
    "postal_town": {},
    "southwest": {},
    "street_number": {},
    "route": {},
    "error_message": "Invalid request. Invalid 'latlng' parameter."
}
{
    "status_code": 200,
    "location": "45.15, -75.14",
    "encoding": "utf-8",
    "provider": "google",
    "status": "INVALID_REQUEST",
    "ok": false
}

OSM Quality
-----------
(0/-2)

Fieldnames
----------
- [ ] accuracy
- [ ] address
- [ ] bbox
- [ ] city
- [ ] client
- [ ] client_secret
- [ ] confidence
- [ ] country
- [ ] county
- [x] encoding
- [ ] housenumber
- [ ] lat
- [ ] lng
- [x] location
- [ ] neighborhood
- [ ] ok
- [ ] place
- [ ] postal
- [ ] postal_town
- [x] provider
- [ ] quality
- [ ] state
- [x] status
- [

In [3]:
location = geolocator.reverse("40.820189,-73.8906752")

NameError: name 'geolocator' is not defined

In [288]:
location.raw['address']['postcode']

'10472'

In [289]:
location.raw

{'address': {'city': 'NYC',
  'country': 'United States of America',
  'country_code': 'us',
  'county': 'Bronx County',
  'neighbourhood': 'Hunts Point',
  'park': 'Del Valle Square',
  'postcode': '10472',
  'road': 'Bruckner Boulevard',
  'state': 'New York',
  'suburb': 'Bronx'},
 'boundingbox': ['40.8204512', '40.8210196', '-73.8914092', '-73.8900306'],
 'display_name': 'Del Valle Square, Bruckner Boulevard, Hunts Point, Bronx, Bronx County, NYC, New York, 10472, United States of America',
 'lat': '40.8206867',
 'licence': 'Data © OpenStreetMap contributors, ODbL 1.0. http://www.openstreetmap.org/copyright',
 'lon': '-73.8907829225018',
 'osm_id': '237040790',
 'osm_type': 'way',
 'place_id': '119465840'}

In [290]:
location.raw['address']['suburb']

'Bronx'

In [13]:
aadt = pd.read_csv('aadt_zip_clean.csv', low_memory=False)

In [210]:
aadt[aadt['BOROUGH'].isnull()]

DATE   TIME BOROUGH ZIP CODE   LATITUDE  LONGITUDE  \
14141   02/28/2017   8:00     NaN      NaN  40.886326 -73.810660   
15880   02/24/2017  14:30     NaN      NaN  40.886127 -73.815320   
15949   02/24/2017  16:05     NaN      NaN  40.886127 -73.815320   
20588   02/15/2017  17:30     NaN      NaN  40.886127 -73.815320   
44782   12/24/2016  13:50     NaN      NaN  40.886127 -73.815320   
50137   12/13/2016   6:15     NaN      NaN  40.908203 -73.886060   
55639   11/30/2016  14:11     NaN      NaN  40.886172 -73.815443   
60476   11/19/2016  22:11     NaN      NaN  40.886133 -73.815316   
63196   11/13/2016  14:11     NaN      NaN  40.886133 -73.815316   
64614   11/10/2016  21:11     NaN      NaN  40.887593 -73.815246   
69865   10/29/2016  20:10     NaN      NaN  40.887593 -73.815246   
75354   10/17/2016  19:10     NaN      NaN  40.887593 -73.815246   
78571   10/09/2016  10:10     NaN      NaN  40.747238 -73.772872   
79466   10/07/2016  15:05     NaN      NaN  40.908226 -73.886040   
82819   09/11/2016  13:09     NaN      NaN  40.748370 -73.895965   
82843   09/11/2016  14:09     NaN      NaN  40.886133 -73.815316   
152710  03/12/2016  11:30     NaN      NaN  40.887593 -73.815246   
175414  01/26/2016  18:20     NaN      NaN  40.887593 -73.815246   
198004  12/11/2015  17:15     NaN      NaN  40.886133 -73.815316   
198020  12/11/2015  17:44     NaN      NaN  40.886133 -73.815316   
217491  11/04/2015   8:15     NaN      NaN  40.886133 -73.815316   
246467  09/10/2015   8:00     NaN      NaN  40.886133 -73.815316   
263622  08/07/2015  13:40     NaN      NaN  40.887593 -73.815246   
266388  08/02/2015  14:10     NaN      NaN  40.886133 -73.815316   
271342  07/24/2015   6:59     NaN      NaN  40.887593 -73.815246   
281031  07/05/2015  15:55     NaN      NaN  40.887593 -73.815246   
285514  06/26/2015  16:45     NaN      NaN  40.887593 -73.815246   
286675  06/24/2015  15:15     NaN      NaN  40.586118 -73.936033   
286691  06/24/2015  15:30     NaN      NaN  40.702617 -73.960572   
286706  06/24/2015  15:58     NaN      NaN  40.623484 -74.019929   
286717  06/24/2015  16:00     NaN      NaN  40.767373 -73.950057   
286738  06/24/2015  16:30     NaN      NaN  40.764651 -73.823484   
286740  06/24/2015  16:30     NaN      NaN  40.792439 -73.934259   
286742  06/24/2015  16:35     NaN      NaN  40.875231 -73.874515   
286743  06/24/2015  16:35     NaN      NaN  40.873644 -73.907962   
286769  06/24/2015  17:00     NaN      NaN  40.747578 -73.761460   
286776  06/24/2015  17:10     NaN      NaN  40.789386 -73.782317   
286793  06/24/2015  18:00     NaN      NaN  40.728858 -73.927810   
286796  06/24/2015  17:40     NaN      NaN  40.823677 -73.872325   
286802  06/24/2015  17:45     NaN      NaN  40.584255 -73.963787   
286822  06/24/2015  18:19     NaN      NaN  40.788493 -73.789716   
286831  06/24/2015  18:28     NaN      NaN  40.758258 -73.777191   
286832  06/24/2015  18:28     NaN      NaN  40.758258 -73.777191   
286836  06/24/2015  18:30     NaN      NaN  40.555540 -74.175612   
286839  06/24/2015  18:33     NaN      NaN  40.857782 -73.830986   
286849  06/24/2015  18:44     NaN      NaN  40.683097 -73.805752   
286873  06/24/2015  19:15     NaN      NaN  40.692311 -73.881456   
286922  06/24/2015  20:30     NaN      NaN  40.739575 -73.791661   
286924  06/24/2015  20:45     NaN      NaN  40.727873 -73.929122   
286926  06/24/2015  21:00     NaN      NaN  40.681739 -73.805091   
...            ...    ...     ...      ...        ...        ...   
799959  07/02/2012  17:00     NaN      NaN  40.749286 -73.756634   
799960  07/02/2012  17:04     NaN      NaN  40.863804 -73.933434   
799975  07/02/2012  17:26     NaN      NaN  40.635808 -74.017614   
799992  07/02/2012  17:49     NaN      NaN  40.831979 -73.935145   
800057  07/02/2012  19:30     NaN      NaN  40.711999 -73.822850   
800110  07/01/2012  17:00     NaN      NaN  40.853709 -73.871857   
800130  07/01/2012   5:00     NaN      NaN  40.537332 -74.189127   
80

In [211]:
aadt[aadt['ZIP CODE'].isnull()]

DATE   TIME BOROUGH ZIP CODE   LATITUDE  LONGITUDE  \
14141   02/28/2017   8:00     NaN      NaN  40.886326 -73.810660   
15880   02/24/2017  14:30     NaN      NaN  40.886127 -73.815320   
15949   02/24/2017  16:05     NaN      NaN  40.886127 -73.815320   
20588   02/15/2017  17:30     NaN      NaN  40.886127 -73.815320   
44782   12/24/2016  13:50     NaN      NaN  40.886127 -73.815320   
50137   12/13/2016   6:15     NaN      NaN  40.908203 -73.886060   
55639   11/30/2016  14:11     NaN      NaN  40.886172 -73.815443   
60476   11/19/2016  22:11     NaN      NaN  40.886133 -73.815316   
63196   11/13/2016  14:11     NaN      NaN  40.886133 -73.815316   
64614   11/10/2016  21:11     NaN      NaN  40.887593 -73.815246   
69865   10/29/2016  20:10     NaN      NaN  40.887593 -73.815246   
75354   10/17/2016  19:10     NaN      NaN  40.887593 -73.815246   
78571   10/09/2016  10:10     NaN      NaN  40.747238 -73.772872   
79466   10/07/2016  15:05     NaN      NaN  40.908226 -73.886040   
82819   09/11/2016  13:09     NaN      NaN  40.748370 -73.895965   
82843   09/11/2016  14:09     NaN      NaN  40.886133 -73.815316   
152710  03/12/2016  11:30     NaN      NaN  40.887593 -73.815246   
175414  01/26/2016  18:20     NaN      NaN  40.887593 -73.815246   
198004  12/11/2015  17:15     NaN      NaN  40.886133 -73.815316   
198020  12/11/2015  17:44     NaN      NaN  40.886133 -73.815316   
217491  11/04/2015   8:15     NaN      NaN  40.886133 -73.815316   
246467  09/10/2015   8:00     NaN      NaN  40.886133 -73.815316   
263622  08/07/2015  13:40     NaN      NaN  40.887593 -73.815246   
266388  08/02/2015  14:10     NaN      NaN  40.886133 -73.815316   
271342  07/24/2015   6:59     NaN      NaN  40.887593 -73.815246   
281031  07/05/2015  15:55     NaN      NaN  40.887593 -73.815246   
285514  06/26/2015  16:45     NaN      NaN  40.887593 -73.815246   
286675  06/24/2015  15:15     NaN      NaN  40.586118 -73.936033   
286691  06/24/2015  15:30     NaN      NaN  40.702617 -73.960572   
286706  06/24/2015  15:58     NaN      NaN  40.623484 -74.019929   
286717  06/24/2015  16:00     NaN      NaN  40.767373 -73.950057   
286738  06/24/2015  16:30     NaN      NaN  40.764651 -73.823484   
286740  06/24/2015  16:30     NaN      NaN  40.792439 -73.934259   
286742  06/24/2015  16:35     NaN      NaN  40.875231 -73.874515   
286743  06/24/2015  16:35     NaN      NaN  40.873644 -73.907962   
286769  06/24/2015  17:00     NaN      NaN  40.747578 -73.761460   
286776  06/24/2015  17:10     NaN      NaN  40.789386 -73.782317   
286793  06/24/2015  18:00     NaN      NaN  40.728858 -73.927810   
286796  06/24/2015  17:40     NaN      NaN  40.823677 -73.872325   
286802  06/24/2015  17:45     NaN      NaN  40.584255 -73.963787   
286822  06/24/2015  18:19     NaN      NaN  40.788493 -73.789716   
286831  06/24/2015  18:28     NaN      NaN  40.758258 -73.777191   
286832  06/24/2015  18:28     NaN      NaN  40.758258 -73.777191   
286836  06/24/2015  18:30     NaN      NaN  40.555540 -74.175612   
286839  06/24/2015  18:33     NaN      NaN  40.857782 -73.830986   
286849  06/24/2015  18:44     NaN      NaN  40.683097 -73.805752   
286873  06/24/2015  19:15     NaN      NaN  40.692311 -73.881456   
286922  06/24/2015  20:30     NaN      NaN  40.739575 -73.791661   
286924  06/24/2015  20:45     NaN      NaN  40.727873 -73.929122   
286926  06/24/2015  21:00     NaN      NaN  40.681739 -73.805091   
...            ...    ...     ...      ...        ...        ...   
799959  07/02/2012  17:00     NaN      NaN  40.749286 -73.756634   
799960  07/02/2012  17:04     NaN      NaN  40.863804 -73.933434   
799975  07/02/2012  17:26     NaN      NaN  40.635808 -74.017614   
799992  07/02/2012  17:49     NaN      NaN  40.831979 -73.935145   
800057  07/02/2012  19:30     NaN      NaN  40.711999 -73.822850   
800110  07/01/2012  17:00     NaN      NaN  40.853709 -73.871857   
800130  07/01/2012   5:00     NaN      NaN  40.537332 -74.189127   
80

# Stopped at 285514 -NR
# Complete up to 318000 (start from here) - MH

In [48]:
for d in range(318000,len(aadt)):
    empty_borough = aadt.at[d,'BOROUGH']
    empty_zip = aadt.at[d,'ZIP CODE']
    if str(empty_borough) == 'nan' or str(empty_zip) == 'nan':
        coordinates = aadt.at[d,'LOCATION']
        location = geolocator.reverse(coordinates[1:-1],timeout=1000)
        zip_code = location.raw['address']
        zip_code = zip_code.get('postcode', 'nan')
        aadt.at[d,'ZIP CODE']=zip_code
        borough = location.raw['address']['county'].split(' ')[0].upper()
        if borough == 'KINGS':
            aadt.at[d,'BOROUGH']= 'BROOKLYN'
        elif borough == 'NEW':
            aadt.at[d,'BOROUGH']= 'MANHATTAN'
        elif borough == 'RICHMOND':
            aadt.at[d,'BOROUGH']= 'STATEN ISLAND'
        else: 
            aadt.at[d,'BOROUGH']= borough

GeocoderServiceError: HTTP Error 429: Too Many Requests

# Before any passes through script:

In [312]:
# aadt['BOROUGH'].isnull()
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

92595
92644


# 1st Run:

In [314]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

91811
91859


# 2nd Run:
dataframe from 0-7000 needs to be re-run or fixed for inconcistent county names

In [317]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

90220
90268


# 3rd Run:

In [324]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

90144
90192


# 4th Run

In [333]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

89882
89930


# 5th Run

In [340]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

89669
89717


# 6th Run

In [344]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

89542
89590


# 7th Run

In [347]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

89494
89542


# 8th Run

In [349]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

89410
89458


# 9th Run

In [358]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

89389
89437


# 10th Run

In [362]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

89341
89389


# 11th Run

In [364]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

89338
89386


# 12th Run

In [379]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

89281
89329


# 13 Run

In [384]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

87775
87822


# 14th Run

In [401]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

87189
87236


# 15th Run

In [406]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

85606
85650


# 16th Run

In [413]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

84158
84201


# 17th Run

In [424]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

82465
82508


# 18th & 19th Runs

In [99]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

74437
74472


# 20th Run

In [113]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

72685
72720


# 21st Run

In [128]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

68158
68192


# 22nd Run

In [134]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

66484
66517


# 23rd Run

In [143]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

61957
61980


# 24th Run

In [154]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

60300
60322


# 25th Run

In [162]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

58617
58639


# 26th Run

In [169]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

56943
56962


# 27th Run

In [176]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

55279
55294


# 28th Run

In [181]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

53569
53582


# 29th Run

In [201]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

49049
49055


# 30th Run

In [15]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

47307
47311


# 31st Run

In [49]:
print (len(aadt[pd.isnull(aadt['BOROUGH'])]))
print (len(aadt[pd.isnull(aadt['ZIP CODE'])]))

45599
45602


In [37]:
aadt[318000:]

DATE   TIME        BOROUGH ZIP CODE   LATITUDE  LONGITUDE  \
318000  04/25/2015  18:00      MANHATTAN  10007.0  40.716467 -74.013080   
318001  04/25/2015  18:02       BROOKLYN  11206.0  40.709601 -73.934872   
318002  04/25/2015  18:05         QUEENS    11374  40.735799 -73.858061   
318003  04/25/2015  18:06          BRONX  10462.0  40.848950 -73.862062   
318004  04/25/2015  18:09      MANHATTAN  10026.0  40.800031 -73.954855   
318005  04/25/2015  18:10         QUEENS  11691.0  40.595805 -73.764646   
318006  04/25/2015  18:10       BROOKLYN  11204.0  40.610846 -73.991767   
318007  04/25/2015  18:15       BROOKLYN  11235.0  40.576304 -73.968531   
318008  04/25/2015  18:18      MANHATTAN  10019.0  40.767110 -73.981718   
318009  04/25/2015  18:20      MANHATTAN  10007.0  40.713646 -74.003604   
318010  04/25/2015  18:20      MANHATTAN  10009.0  40.729157 -73.975271   
318011  04/25/2015  18:20      MANHATTAN  10036.0  40.762127 -73.997387   
318012  04/25/2015  18:20       BROOKLYN  11236.0  40.632206 -73.909599   
318013  04/25/2015  18:20      MANHATTAN  10016.0  40.744498 -73.981108   
318014  04/25/2015  18:22       BROOKLYN  11216.0  40.678320 -73.946919   
318015  04/25/2015  18:25       BROOKLYN  11206.0  40.710094 -73.937209   
318016  04/25/2015  18:30       BROOKLYN  11212.0  40.662833 -73.926138   
318017  04/25/2015  18:30      MANHATTAN  10036.0  40.759730 -73.991683   
318018  04/25/2015  18:30         QUEENS  11419.0  40.688293 -73.831853   
318019  04/25/2015  18:30         QUEENS  11435.0  40.714935 -73.816354   
318020  04/25/2015  18:30         QUEENS    11105  40.768881 -73.908515   
318021  04/25/2015  18:40      MANHATTAN  10036.0  40.759790 -73.987926   
318022  04/25/2015  18:40         QUEENS  11373.0  40.741289 -73.885727   
318023  04/25/2015  18:40         QUEENS  11373.0  40.741289 -73.885727   
318024  04/25/2015  18:45         QUEENS  11412.0  40.700567 -73.764100   
318025  04/25/2015  18:50      MANHATTAN  10016.0  40.745441 -73.975394   
318026  04/25/2015  18:52      MANHATTAN  10019.0  40.759842 -73.984175   
318027  04/25/2015  18:55      MANHATTAN  10019.0  40.762966 -73.973969   
318028  04/25/2015  18:55       BROOKLYN  11220.0  40.642058 -74.020420   
318029  04/25/2015  19:00       BROOKLYN  11214.0  40.615710 -74.001448   
318030  04/25/2015  19:00      MANHATTAN  10036.0  40.759730 -73.991683   
318031  04/25/2015  19:00         QUEENS  11102.0  40.770411 -73.934474   
318032  04/25/2015  19:03          BRONX  10467.0  40.867804 -73.862504   
318033  04/25/2015  19:10         QUEENS  11378.0  40.732828 -73.918869   
318034  04/25/2015  19:10       BROOKLYN    11234  40.623572 -73.935612   
318035  04/25/2015  19:10       BROOKLYN  11207.0  40.670937 -73.902290   
318036  04/25/2015  19:15         QUEENS  11356.0  40.781684 -73.845938   
318037  04/25/2015  19:15          BRONX  10457.0  40.854069 -73.899317   
318038  04/25/2015  19:20         QUEENS  11369.0  40.763009 -73.875322   
318039  04/25/2015  19:28         QUEENS    11368  40.738416 -73.848984   
318040  04/25/2015  19:30      MANHATTAN  10001.0  40.751523 -73.993947   
318041  04/25/2015  19:30      MANHATTAN  10032.0  40.842438 -73.939253   
318042  04/25/2015  19:40         QUEENS  11101.0  40.752976 -73.910233   
318043  04/25/2015  19:40         QUEENS  11365.0  40.740493 -73.807396   
318044  04/25/2015  19:41       BROOKLYN  11231.0  40.673471 -73.998389   
318045  04/25/2015  19:45         QUEENS  11436.0  40.667977 -73.798132   
318046  04/25/2015  19:50       BROOKLYN  11208.0  40.660756 -73.876101   
318047  04/25/2015  19:51         QUEENS  11379.0  40.712144 -73.873528   
318048  04/25/2015   1:00       BROOKLYN  11232.0  40.660562 -74.001189   
318049  04/25/2015   1:00      MANHATTAN    10128  40.778872 -73.943235   
...            ...    ...            ...      ...        ...        ...   
800628  07/01/2012   5:50      MANHATTAN    10019  40.765365 -73.987578   
800629  07/01/2012   6:15         QUEEN

In [10]:
location = geolocator.reverse('40.692158,-73.999115')

In [11]:
zip_code = location.raw['address']
zip_code = zip_code.get('postcode', 'nan')
zip_code

'11217'

In [50]:
aadt.to_csv('aadt_zip_clean2.csv', index=False)